File: shared/src/commonMain/kotlin/com/vnidrop/app/core/CoreRepository.kt
package com.vnidrop.app.core

import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.IO
import kotlinx.coroutines.channels.BufferOverflow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.withContext
import kotlin.random.Random
import uniffi.vnidrop.CoreEvent
import uniffi.vnidrop.CoreEventSink
import uniffi.vnidrop.CoreNetworkConfig
import uniffi.vnidrop.CoreRelayMode
import uniffi.vnidrop.ReceiveOutputSink
import uniffi.vnidrop.ReceiveOutputSinkV2
import uniffi.vnidrop.ReceiverRequest
import uniffi.vnidrop.ShareMetadataInput
import uniffi.vnidrop.ShareResult
import uniffi.vnidrop.ShareSource
import uniffi.vnidrop.SourceKind
import uniffi.vnidrop.StoredTransfer
import uniffi.vnidrop.TicketInspection
import uniffi.vnidrop.TransferMetadata
import uniffi.vnidrop.TransferAccessMode
import uniffi.vnidrop.VnidropCore
import uniffi.vnidrop.clearInactiveTransferCache
import uniffi.vnidrop.defaultCoreNetworkConfig

class CoreRepository(
	private val dispatcher: CoroutineDispatcher = Dispatchers.IO,
) : CoreGateway {
	private val _state = MutableStateFlow(CoreState())
	override val state: StateFlow<CoreState> = _state.asStateFlow()

	private val _signals = MutableSharedFlow<CoreSignal>(
		extraBufferCapacity = 64,
		onBufferOverflow = BufferOverflow.DROP_OLDEST,
	)
	override val signals: SharedFlow<CoreSignal> = _signals.asSharedFlow()

	private var core: VnidropCore? = null
	private var currentAppDataDir: String? = null
	private var currentRelaySettings = RelaySettings()
	private val lifecycleGate = CoreLifecycleGate()

	private val sink = object : CoreEventSink {
		override fun onEvent(event: CoreEvent) {
			val model = event.toModel()
			_state.update { current -> current.copy(events = (listOf(model) + current.events).take(MaxEvents)) }
			val transferId = model.transferId
			if (transferId != null) {
				when (model.phase) {
					"approval", "access" -> _signals.tryEmit(CoreSignal.ApprovalChanged(transferId))
					"delivery" -> _signals.tryEmit(CoreSignal.ReceiverHistoryChanged(transferId))
				}
				if (model.shouldRefreshTransfers()) {
					_signals.tryEmit(CoreSignal.TransfersChanged(transferId))
				}
			}
		}
	}

	override suspend fun initialize(appDataDir: String, relaySettings: RelaySettings): Result<Unit> =
		runReconfiguration {
			val previousCore = core
			if (previousCore != null) {
				val status = previousCore.status()
				require(status.activeTransfers == 0UL && status.activeShares == 0UL) {
					"Stop active transfers and shares before changing relay settings"
				}
			}
			_state.update { it.copy(isInitialized = false, status = null) }
			core = null
			disposeCore(previousCore)
			core = VnidropCore.initializeWithNetworkConfig(appDataDir, sink, relaySettings.toNative())
			currentAppDataDir = appDataDir
			currentRelaySettings = relaySettings
			refreshSnapshot(requireCore())
			_state.update { it.copy(isInitialized = true) }
		}

	override fun shutdown() {
		val activeCore = core
		core = null
		disposeCore(activeCore)
		currentAppDataDir = null
		_state.value = CoreState()
	}

	override suspend fun sharePath(
		path: String,
		transferName: String,
		senderName: String,
		accessPolicy: ShareAccessPolicy,
	): Result<Share> =
		shareSources(
			sources = listOf(
				ShareSource(
					kind = SourceKind.PATH,
					value = path,
					displayName = path.substringAfterLast('/').ifBlank { null },
					isDirectory = false,
				),
			),
			transferName = transferName,
			senderName = senderName,
			accessPolicy = accessPolicy,
		)

	override suspend fun shareFileDescriptor(
		fd: Int,
		displayName: String,
		transferName: String,
		senderName: String,
		accessPolicy: ShareAccessPolicy,
	): Result<Share> =
		shareSources(
			sources = listOf(
				ShareSource(
					kind = SourceKind.FILE_DESCRIPTOR,
					value = fd.toString(),
					displayName = displayName.ifBlank { "transfer" },
					isDirectory = false,
				),
			),
			transferName = transferName,
			senderName = senderName,
			accessPolicy = accessPolicy,
		)

	override suspend fun inspectTicket(ticket: String): Result<TicketInspectionModel> = runCore { activeCore ->
		activeCore.inspectTicket(ticket).toModel().also { inspection ->
			_state.update { it.copy(lastInspection = inspection) }
		}
	}

	override suspend fun receive(ticket: String, outputDir: String, receiverName: String): Result<Unit> = runCore { activeCore ->
		activeCore.receive(ticket, outputDir, receiverName.ifBlank { null })
		refreshSnapshot(activeCore)
	}

	override suspend fun receiveWithOutputSink(
		ticket: String,
		outputSink: ReceiveOutputSink,
		receiverName: String,
	): Result<Unit> = runCore { activeCore ->
		activeCore.receiveWithOutputSink(ticket, outputSink, receiverName.ifBlank { null })
		refreshSnapshot(activeCore)
	}

	override suspend fun receiveWithOutputSinkV2(
		ticket: String,
		outputSink: ReceiveOutputSinkV2,
		receiverName: String,
	): Result<Unit> = runCore { activeCore ->
		activeCore.receiveWithOutputSinkV2(ticket, outputSink, receiverName.ifBlank { null })
		refreshSnapshot(activeCore)
	}

	override suspend fun storageUsage(): Result<CoreStorageUsageModel> = runCore { activeCore ->
		val usage = activeCore.storageUsage()
		CoreStorageUsageModel(
			blobStoreBytes = usage.blobStoreBytes,
			databaseBytes = usage.databaseBytes,
			logsBytes = usage.logsBytes,
			previewsBytes = usage.previewsBytes,
			otherCoreBytes = usage.otherCoreBytes,
		)
	}

	override suspend fun clearTransferCache(): Result<ULong> = runReconfiguration {
		val activeCore = requireCore()
		val status = activeCore.status()
		require(status.activeTransfers == 0UL && status.activeShares == 0UL) {
			"Wait for active transfers and shares to finish before clearing the transfer cache"
		}
		val appDataDir = requireNotNull(currentAppDataDir) { "Core data directory is unavailable" }
		val relaySettings = currentRelaySettings
		_state.update { it.copy(isInitialized = false, status = null) }
		core = null
		disposeCore(activeCore)
		var reclaimed = 0UL
		try {
			reclaimed = clearInactiveTransferCache(appDataDir)
		} finally {
			core = VnidropCore.initializeWithNetworkConfig(appDataDir, sink, relaySettings.toNative())
			refreshSnapshot(requireCore())
			_state.update { it.copy(isInitialized = true) }
		}
		reclaimed
	}

	private fun disposeCore(activeCore: VnidropCore?) {
		if (activeCore == null) return
		try {
			activeCore.shutdown()
		} finally {
			// UniFFI owns the Rust Arc; explicit disposal releases Windows file handles before cache deletion.
			activeCore.close()
		}
	}

	override suspend fun receivedArtifacts(): Result<List<ReceivedArtifactModel>> = runCore { activeCore ->
		activeCore.listReceivedArtifacts().map { artifact ->
			ReceivedArtifactModel(
				id = artifact.id,
				locator = artifact.locator,
				locatorKind = artifact.locatorKind,
				logicalSize = artifact.logicalSize,
			)
		}
	}

	override suspend fun cancel(transferId: ULong): Result<Unit> = runCore { activeCore ->
		activeCore.cancelTransfer(transferId)
		refreshSnapshot(activeCore)
	}

	override suspend fun delete(transferId: ULong): Result<Unit> = runCore { activeCore ->
		activeCore.deleteTransfer(transferId)
		refreshSnapshot(activeCore)
		_signals.tryEmit(CoreSignal.ApprovalChanged(transferId))
		_signals.tryEmit(CoreSignal.ReceiverHistoryChanged(transferId))
	}

	override suspend fun clearReceiveHistory(): Result<ULong> = runCore { activeCore ->
		val deleted = activeCore.deleteReceiveHistory()
		refreshSnapshot(activeCore)
		deleted
	}

	override suspend fun receiverRequests(transferId: ULong): Result<List<ReceiverRequestModel>> = runCore { activeCore ->
		activeCore.listReceiverRequests(transferId).map(ReceiverRequest::toModel)
	}

	override suspend fun respondReceiverRequest(
		requestId: String,
		accepted: Boolean,
		reason: String?,
	): Result<Unit> = runCore { activeCore ->
		activeCore.respondReceiverRequest(requestId, accepted, reason)
	}

	override suspend fun refresh(): Result<Unit> = runCore(::refreshSnapshot)

	override suspend fun shareSources(
		sources: List<ShareSource>,
		transferName: String,
		senderName: String,
		accessPolicy: ShareAccessPolicy,
	): Result<Share> = runCore { activeCore ->
		require(sources.isNotEmpty()) { "Select at least one file to share" }
		withPlatformPathAccess(sources) {
			activeCore.shareFiles(
				sources = sources,
				metadata = ShareMetadataInput(
					transferId = nextTransferId(),
					transferName = transferName.ifBlank { null },
					senderName = senderName.ifBlank { null },
					accessMode = accessPolicy.toNative(),
				),
			).toModel()
		}.also { share ->
			refreshSnapshot(activeCore)
			_state.update { it.copy(lastShare = share) }
		}
	}

	private suspend fun <T> withPlatformPathAccess(
		sources: List<ShareSource>,
		index: Int = 0,
		block: suspend () -> T,
	): T {
		if (index >= sources.size) return block()
		val source = sources[index]
		return withPlatformPathAccess(source.kind, source.value) {
			withPlatformPathAccess(sources, index + 1, block)
		}
	}

	private fun refreshSnapshot(activeCore: VnidropCore) {
		val status = activeCore.status()
		_state.update {
			it.copy(
				status = CoreStatus(status.endpointId, status.activeTransfers, status.activeShares),
				transfers = activeCore.listTransfers().map(StoredTransfer::toModel),
				events = activeCore.listEvents(null).map(CoreEvent::toModel).take(MaxEvents),
			)
		}
	}

	private suspend fun <T> runCore(block: suspend (VnidropCore) -> T): Result<T> =
		withContext(dispatcher) {
			try {
				Result.success(
					lifecycleGate.withCall(
						capture = ::requireCore,
						block = block,
					),
				)
			} catch (error: Throwable) {
				if (error is CancellationException) throw error
				Result.failure(error)
			}
		}

	private suspend fun <T> runReconfiguration(block: suspend () -> T): Result<T> =
		withContext(dispatcher) {
			try {
				Result.success(lifecycleGate.withReconfiguration(block))
			} catch (error: Throwable) {
				if (error is CancellationException) throw error
				Result.failure(error)
			}
		}

	private fun requireCore(): VnidropCore = core ?: error("Initialize the core first.")

	private fun nextTransferId(): ULong = Random.nextLong(1, Long.MAX_VALUE).toULong()

	private companion object {
		const val MaxEvents = 200
	}
}

private fun RelaySettings.toNative(): CoreNetworkConfig = when (mode) {
	RelayMode.Automatic -> defaultCoreNetworkConfig()
	RelayMode.StrictCustom -> CoreNetworkConfig(
		mode = CoreRelayMode.STRICT_CUSTOM,
		relayUrls = relayUrls,
	)
	RelayMode.CustomWithDirectFallback -> CoreNetworkConfig(
		mode = CoreRelayMode.CUSTOM_WITH_DIRECT_FALLBACK,
		relayUrls = relayUrls,
	)
	RelayMode.LocalOnly -> CoreNetworkConfig(
		mode = CoreRelayMode.LOCAL_ONLY,
		relayUrls = emptyList(),
	)
}

private fun CoreEvent.toModel(): CoreEventModel = CoreEventModel(
	id = id,
	timestamp = timestamp,
	scope = scope,
	transferId = transferId,
	direction = direction,
	phase = phase,
	kind = kind,
	dataJson = dataJson,
)

private fun CoreEventModel.shouldRefreshTransfers(): Boolean =
	phase in setOf("lifecycle", "error", "ticket", "import", "download", "export", "handshake") &&
		kind in setOf(
			"started", "done", "created", "failed", "cancelled", "share-stopped",
			"found-collection", "connected",
		)

private fun StoredTransfer.toModel(): Transfer = Transfer(
	localId = localId,
	transferId = transferId,
	direction = direction.toTransferDirection(),
	status = status.toTransferStatus(),
	peerId = peerId,
	transferName = transferName,
	contentHash = contentHash,
	fileCount = fileCount,
	totalSize = totalSize,
	ticket = ticket,
	accessPolicy = accessMode.toModel(),
	createdAt = createdAt,
	updatedAt = updatedAt,
)

private fun ShareAccessPolicy.toNative(): TransferAccessMode = when (this) {
	ShareAccessPolicy.RequireApproval -> TransferAccessMode.APPROVAL_REQUIRED
	ShareAccessPolicy.AnyoneWithTransfer -> TransferAccessMode.PUBLIC
}

private fun TransferAccessMode.toModel(): ShareAccessPolicy = when (this) {
	TransferAccessMode.APPROVAL_REQUIRED -> ShareAccessPolicy.RequireApproval
	TransferAccessMode.PUBLIC -> ShareAccessPolicy.AnyoneWithTransfer
}

private fun String.toTransferDirection(): TransferDirection = when (this) {
	"send" -> TransferDirection.Send
	"receive" -> TransferDirection.Receive
	else -> error("Unknown transfer direction: $this")
}

private fun String.toTransferStatus(): TransferStatus = when (this) {
	"importing" -> TransferStatus.Importing
	"sharing" -> TransferStatus.Sharing
	"receiving" -> TransferStatus.Receiving
	"done" -> TransferStatus.Done
	"failed" -> TransferStatus.Failed
	"cancelled" -> TransferStatus.Cancelled
	"stopped" -> TransferStatus.Stopped
	else -> error("Unknown transfer status: $this")
}

private fun ShareResult.toModel(): Share = Share(
	transferId = transferId,
	ticket = ticket,
	transferName = transferName,
	contentHash = hash,
	fileCount = fileCount,
	totalSize = totalSize,
)

private fun TicketInspection.toModel(): TicketInspectionModel = TicketInspectionModel(
	kind = kind,
	metadata = metadata.toModel(),
)

private fun TransferMetadata.toModel(): TransferMetadataModel = TransferMetadataModel(
	transferId = transferId,
	transferName = transferName,
	senderName = senderName,
	contentHash = contentHash,
	fileCount = fileCount,
	totalSize = totalSize,
)

private fun ReceiverRequest.toModel(): ReceiverRequestModel = ReceiverRequestModel(
	id = id,
	transferId = transferId,
	remoteEndpointId = remoteEndpointId,
	transferName = transferName,
	receiverName = receiverName,
	receiverDeviceName = receiverDeviceName,
	appVersion = appVersion,
	status = when (status) {
		"requested" -> ReceiverDeliveryStatus.Requested
		"accepted" -> ReceiverDeliveryStatus.Accepted
		"refused" -> ReceiverDeliveryStatus.Refused
		"expired" -> ReceiverDeliveryStatus.Expired
		"completed" -> ReceiverDeliveryStatus.Completed
		"failed" -> ReceiverDeliveryStatus.Failed
		else -> ReceiverDeliveryStatus.Unknown
	},
	reason = reason,
	requestedAt = requestedAt,
	respondedAt = respondedAt,
	completedAt = completedAt,
)

File: apple/VniDrop/Core/CoreRepository.swift
import Foundation
import Combine
@preconcurrency import VnidropCore

/// Swift port of `core/CoreRepository.kt`. Owns the `VnidropCore` handle, maps the
/// generated UniFFI records into app domain models, publishes an observable
/// `CoreState`, and emits coalesced `CoreSignal`s from the event sink.
///
/// UniFFI calls block (the core drives its own runtime via `block_on`), so they
/// run on a background queue and results are hopped back to the main actor.
@MainActor
final class CoreRepository: ObservableObject, CoreGateway {
	@Published private(set) var state = CoreState()
	var statePublisher: AnyPublisher<CoreState, Never> { $state.eraseToAnyPublisher() }

	private let signalsSubject = PassthroughSubject<CoreSignal, Never>()
	/// Coalesced change hints; subscribe to react to approval/history/transfer changes.
	var signals: AnyPublisher<CoreSignal, Never> { signalsSubject.eraseToAnyPublisher() }

	// Set on the main actor (initialize/shutdown) but read from `queue` inside
	// `runCore`; the underlying core is internally synchronized, so this crossing
	// is safe. `nonisolated(unsafe)` documents that contract for Swift 6.
	private nonisolated(unsafe) var core: VnidropCore?
	private let dispatcher = CoreDispatcher()
	private lazy var sink = RepositoryEventSink { [weak self] event in
		Task { @MainActor in self?.handle(event: event) }
	}

	private nonisolated static let maxEvents = 200

	// MARK: - Lifecycle

	func initialize(appDataDir: String) async -> Result<Void, Error> {
		await runCore { [sink] in
			self.core?.shutdown()
			let created = try VnidropCore.initialize(appDataDir: appDataDir, eventSink: sink)
			return created
		}.map { created in
			self.core = created
			self.refreshSnapshot()
			self.state.isInitialized = true
		}
	}

	func shutdown() {
		core?.shutdown()
		core = nil
		state = CoreState()
	}

	// MARK: - Share

	func shareSources(
		_ sources: [ShareSource],
		transferName: String,
		senderName: String,
		accessPolicy: ShareAccessPolicy
	) async -> Result<Share, Error> {
		guard !sources.isEmpty else {
			return .failure(InvitationError.message("Select at least one file to share"))
		}
		return await runCore {
			let result = try self.requireCore().shareFiles(
				sources: sources,
				metadata: ShareMetadataInput(
					transferId: Self.nextTransferId(),
					transferName: transferName.isEmpty ? nil : transferName,
					senderName: senderName.isEmpty ? nil : senderName,
					accessMode: accessPolicy.toNative()
				)
			)
			return result.toModel()
		}.map { share in
			self.refreshSnapshot()
			self.state.lastShare = share
			return share
		}
	}

	// MARK: - Inspect / Receive

	func inspectTicket(_ ticket: String) async -> Result<TicketInspectionModel, Error> {
		await runCore {
			try self.requireCore().inspectTicket(ticket: ticket).toModel()
		}.map { inspection in
			self.state.lastInspection = inspection
			return inspection
		}
	}

	func receive(ticket: String, outputDir: String, receiverName: String) async -> Result<Void, Error> {
		await runCore {
			try self.requireCore().receive(
				ticket: ticket,
				outputDir: outputDir,
				receiverName: receiverName.isEmpty ? nil : receiverName
			)
		}.map { self.refreshSnapshot() }
	}

	/// Receive into a security-scoped directory URL, holding access while the core
	/// streams (mirrors `receiveIntoSecurityScopedDirectory`).
	func receiveIntoSecurityScopedDirectory(
		ticket: String,
		outputDirectoryUrl: String,
		receiverName: String
	) async -> Result<Void, Error> {
		await runCore {
			try withSecurityScopedAccess(pathOrUrl: outputDirectoryUrl) {
				try self.requireCore().receive(
					ticket: ticket,
					outputDir: outputDirectoryUrl,
					receiverName: receiverName.isEmpty ? nil : receiverName
				)
			}
		}.map { self.refreshSnapshot() }
	}

	// MARK: - Lifecycle actions

	func cancel(transferId: UInt64) async -> Result<Void, Error> {
		// Off the serial `queue`: a receive in flight is blocking it, and the
		// cancel signal must reach the core to unblock that receive.
		await runInterrupt {
			try self.requireCore().cancelTransfer(transferId: transferId)
		}.map { self.refreshSnapshot() }
	}

	func delete(transferId: UInt64) async -> Result<Void, Error> {
		await runCore {
			try self.requireCore().deleteTransfer(transferId: transferId)
		}.map {
			self.refreshSnapshot()
			self.signalsSubject.send(.approvalChanged(transferId: transferId))
			self.signalsSubject.send(.receiverHistoryChanged(transferId: transferId))
		}
	}

	func clearReceiveHistory() async -> Result<UInt64, Error> {
		await runCore {
			try self.requireCore().deleteReceiveHistory()
		}.map { deleted in
			self.refreshSnapshot()
			return deleted
		}
	}

	func storageUsage() async -> Result<CoreStorageUsageModel, Error> {
		await runCore {
			let usage = try self.requireCore().storageUsage()
			return CoreStorageUsageModel(
				blobStoreBytes: usage.blobStoreBytes,
				appDataBytes: usage.databaseBytes + usage.logsBytes + usage.previewsBytes + usage.otherCoreBytes
			)
		}
	}

	func receivedArtifacts() async -> Result<[ReceivedArtifactModel], Error> {
		await runCore {
			try self.requireCore().listReceivedArtifacts().compactMap { artifact in
				guard artifact.locatorKind == .filesystemPath else { return nil }
				return ReceivedArtifactModel(locator: artifact.locator, logicalSize: artifact.logicalSize)
			}
		}
	}

	func receiverRequests(transferId: UInt64) async -> Result<[ReceiverRequestModel], Error> {
		await runCore {
			try self.requireCore().listReceiverRequests(transferId: transferId).map { $0.toModel() }
		}
	}

	func respondReceiverRequest(
		requestId: String,
		accepted: Bool,
		reason: String? = nil
	) async -> Result<Void, Error> {
		await runCore {
			try self.requireCore().respondReceiverRequest(requestId: requestId, accepted: accepted, reason: reason)
		}
	}

	func refresh() async -> Result<Void, Error> {
		// Read from the core off the main actor, then apply the snapshot on the
		// main actor so `@Published state` is never mutated from `queue`.
		await runCore { self.readSnapshot() }.map { snapshot in
			if let snapshot { self.applySnapshot(snapshot) }
		}
	}

	// MARK: - Event sink handling (ported from CoreRepository.sink)

	private func handle(event: CoreEvent) {
		let model = event.toModel()
		var events = state.events
		events.insert(model, at: 0)
		if events.count > Self.maxEvents { events = Array(events.prefix(Self.maxEvents)) }
		state.events = events

		guard let transferId = model.transferId else { return }
		switch model.phase {
		case "approval", "access": signalsSubject.send(.approvalChanged(transferId: transferId))
		case "delivery": signalsSubject.send(.receiverHistoryChanged(transferId: transferId))
		default: break
		}
		if model.shouldRefreshTransfers {
			signalsSubject.send(.transfersChanged(transferId: transferId))
		}
	}

	// MARK: - Internals

	/// Snapshot of the values read from the core in one pass.
	private struct CoreSnapshot: Sendable {
		let status: CoreStatus
		let transfers: [Transfer]
		let events: [CoreEventModel]
	}

	/// Reads the current core state. Safe to call off the main actor (pure core
	/// FFI reads); does not touch `@Published` state.
	private nonisolated func readSnapshot() -> CoreSnapshot? {
		guard let core = self.core else { return nil }
		let status = core.status()
		let transfers = (try? core.listTransfers())?.map { $0.toModel() } ?? []
		let events = (try? core.listEvents(transferId: nil))?.prefix(Self.maxEvents).map { $0.toModel() } ?? []
		return CoreSnapshot(
			status: CoreStatus(
				endpointId: status.endpointId,
				activeTransfers: status.activeTransfers,
				activeShares: status.activeShares
			),
			transfers: transfers,
			events: Array(events)
		)
	}

	/// Applies a snapshot to `@Published state`. Must run on the main actor.
	private func applySnapshot(_ snapshot: CoreSnapshot) {
		state.status = snapshot.status
		state.transfers = snapshot.transfers
		state.events = snapshot.events
	}

	private func refreshSnapshot() {
		if let snapshot = readSnapshot() { applySnapshot(snapshot) }
	}

	private nonisolated func requireCore() throws -> VnidropCore {
		guard let core = self.core else {
			throw InvitationError.message("Initialize the core first.")
		}
		return core
	}

	/// Runs a blocking core call off the main actor and hops the result back.
	private nonisolated func runCore<T: Sendable>(_ block: @escaping @Sendable () throws -> T) async -> Result<T, Error> {
		await dispatcher.run(block)
	}

	/// Like `runCore`, but off the serial lane so it can interrupt a blocking
	/// call in flight there (e.g. cancel a `receive`). Only use for core calls
	/// that are safe to run concurrently with another core call.
	private nonisolated func runInterrupt<T: Sendable>(_ block: @escaping @Sendable () throws -> T) async -> Result<T, Error> {
		await dispatcher.runInterrupt(block)
	}

	private nonisolated static func nextTransferId() -> UInt64 {
		UInt64.random(in: 1...UInt64(Int64.max))
	}
}

/// Event sink bridged to the repository. `onEvent` is invoked on core-owned
/// threads; the handler hops to the main actor.
private final class RepositoryEventSink: CoreEventSink, @unchecked Sendable {
	private let handler: @Sendable (CoreEvent) -> Void
	init(handler: @escaping @Sendable (CoreEvent) -> Void) { self.handler = handler }
	func onEvent(event: CoreEvent) { handler(event) }
}

/// Runs `body` while holding security-scoped access to a bookmarked URL/path.
private func withSecurityScopedAccess<T>(pathOrUrl: String, _ body: () throws -> T) throws -> T {
	let url = URL(string: pathOrUrl) ?? URL(fileURLWithPath: pathOrUrl)
	let started = url.startAccessingSecurityScopedResource()
	defer { if started { url.stopAccessingSecurityScopedResource() } }
	return try body()
}

// MARK: - Mapping (ported from CoreRepository.kt)

private extension CoreEvent {
	func toModel() -> CoreEventModel {
		CoreEventModel(
			id: id, timestamp: timestamp, scope: scope, transferId: transferId,
			direction: direction, phase: phase, kind: kind, dataJson: dataJson
		)
	}
}

private let refreshPhases: Set<EventPhase> = [.lifecycle, .error, .ticket, .importing, .download, .export, .handshake]
private let refreshKinds: Set<EventKind> = [
	.started, .done, .created, .failed, .cancelled, .shareStopped, .foundCollection, .connected,
]

private extension CoreEventModel {
	var shouldRefreshTransfers: Bool {
		guard let eventPhase, let eventKind else { return false }
		return refreshPhases.contains(eventPhase) && refreshKinds.contains(eventKind)
	}
}

extension ShareAccessPolicy {
	func toNative() -> TransferAccessMode {
		switch self {
		case .requireApproval: return .approvalRequired
		case .anyoneWithTransfer: return .public
		}
	}
}

private extension TransferAccessMode {
	func toModel() -> ShareAccessPolicy {
		switch self {
		case .approvalRequired: return .requireApproval
		case .public: return .anyoneWithTransfer
		}
	}
}

private extension StoredTransfer {
	func toModel() -> Transfer {
		Transfer(
			localId: localId,
			transferId: transferId,
			direction: Self.direction(direction),
			status: Self.status(status),
			peerId: peerId,
			transferName: transferName,
			contentHash: contentHash,
			fileCount: fileCount,
			totalSize: totalSize,
			ticket: ticket,
			accessPolicy: accessMode.toModel(),
			createdAt: createdAt,
			updatedAt: updatedAt
		)
	}

	static func direction(_ raw: String) -> TransferDirection {
		switch raw {
		case "send": return .send
		case "receive": return .receive
		default: return .send
		}
	}

	static func status(_ raw: String) -> TransferStatus {
		switch raw {
		case "importing": return .importing
		case "sharing": return .sharing
		case "receiving": return .receiving
		case "done": return .done
		case "failed": return .failed
		case "cancelled": return .cancelled
		case "stopped": return .stopped
		default: return .failed
		}
	}
}

private extension ShareResult {
	func toModel() -> Share {
		Share(
			transferId: transferId, ticket: ticket, transferName: transferName,
			contentHash: hash, fileCount: fileCount, totalSize: totalSize
		)
	}
}

private extension TicketInspection {
	func toModel() -> TicketInspectionModel {
		TicketInspectionModel(kind: kind, metadata: metadata.toModel())
	}
}

private extension TransferMetadata {
	func toModel() -> TransferMetadataModel {
		TransferMetadataModel(
			transferId: transferId, transferName: transferName, senderName: senderName,
			contentHash: contentHash, fileCount: fileCount, totalSize: totalSize
		)
	}
}

private extension ReceiverRequest {
	func toModel() -> ReceiverRequestModel {
		ReceiverRequestModel(
			id: id, transferId: transferId, remoteEndpointId: remoteEndpointId,
			transferName: transferName, receiverName: receiverName, receiverDeviceName: receiverDeviceName,
			appVersion: appVersion, status: Self.status(status), reason: reason,
			requestedAt: requestedAt, respondedAt: respondedAt, completedAt: completedAt
		)
	}

	static func status(_ raw: String) -> ReceiverDeliveryStatus {
		switch raw {
		case "requested": return .requested
		case "accepted": return .accepted
		case "refused": return .refused
		case "expired": return .expired
		case "completed": return .completed
		default: return .unknown
		}
	}
}
